File: javascript/103.binary-tree-zigzag-level-order-traversal.js
/*
 * @lc app=leetcode id=103 lang=javascript
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
  // bfs variant
  // store node and level with flip switch
  // O(n) time
  // O(n) space
  if (!root) return []
  let queue = []
  let traversalOrder = []
  queue.unshift([root, 0])
  while (queue.length > 0) {
    let vals = queue.shift()
    let currNode = vals[0]
    let currLevel = vals[1]
    if (traversalOrder[currLevel]) {
      traversalOrder[currLevel].push(currNode.val)
    } else {
      traversalOrder[currLevel] = [currNode.val]
    }
    if (currNode.left) {
      queue.push([currNode.left, currLevel+1])
    }
    if (currNode.right) {
      queue.push([currNode.right, currLevel+1])
    }
  }
  // zig-zag ordering
  return traversalOrder.map((v, k) => {
    if (k % 2 === 1) {
      return v.reverse()
    }
    return v
  })
};
// @lc code=end

File: cpp/103.binary-tree-zigzag-level-order-traversal.cpp
/*
 * @lc app=leetcode id=103 lang=cpp
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    struct QueueVal
    {
        TreeNode *root;
        int level;
    };
    vector<vector<int>> zigzagLevelOrder(TreeNode *root)
    {
        // BFS traversal
        // keeping track of order
        // flip direction from left<->right
        // based on even/odd
        // O(n) time O(n) space
        vector<vector<int>> resVect;
        if (!root)
            return resVect;
        queue<QueueVal> myQueue;
        myQueue.push(QueueVal{root, 0});
        while (myQueue.size() > 0)
        {
            QueueVal vals = myQueue.front();
            TreeNode *currNode = vals.root;
            int currLevel = vals.level;
            myQueue.pop();
            if (currLevel == resVect.size())
            {
                // if no vect present at current level
                vector<int> newVect{currNode->val};
                resVect.push_back(newVect);
            }
            else
            {
                resVect[currLevel].push_back(currNode->val);
            }

            if (currNode->left)
            {
                myQueue.push(QueueVal{currNode->left, currLevel + 1});
            }
            if (currNode->right)
            {
                myQueue.push(QueueVal{currNode->right, currLevel + 1});
            }
        }

        // bad way to flip alternating levels
        for (int i = 0; i < resVect.size(); i++)
        {
            if (i % 2 == 1)
            {
                reverse(resVect[i].begin(), resVect[i].end());
            }
        }

        return resVect;
    }
};
// @lc code=end
